src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs

```diff
@@ -3985,6 +3985,1041 @@ .maxstack  2
+
+            compVerifier.VerifyDiagnostics(
+                // (6,25): error CS9365: The given expression cannot be used in a fixed statement
+                //         fixed (int* p = new Fixable(1))
+                Diagnostic(ErrorCode.ERR_ExprCannotBeFixed, "new Fixable(1)").WithLocation(6, 25),
+                // (6,25): error CS0176: Member 'Fixable.DangerousGetPinnableReference()' cannot be accessed with an instance reference; qualify it with a type name instead
+                //         fixed (int* p = new Fixable(1))
+                Diagnostic(ErrorCode.ERR_ObjectProhibited, "new Fixable(1)").WithArguments("Fixable.DangerousGetPinnableReference()").WithLocation(6, 25)
+                );
+        }
+
+        [Fact]
+        public void CustomFixedStructVariousErr03()
+        {
+            var text = @"
+unsafe class C
+{
+    public static void Main()
+    {
+        fixed (int* p = new Fixable(1))
+        {
+            System.Console.Write(p[1]);
+        }
+    }
+}
+
+public struct Fixable
+{
+    public Fixable(int arg){}
+
+    public ref int DangerousGetPinnableReference => ref (new int[]{1,2,3})[0];
+}
+
+";
+
+            var compVerifier = CreateCompilationWithMscorlib46(text, options: TestOptions.UnsafeReleaseExe);
+
+            compVerifier.VerifyDiagnostics(
+                // (6,25): error CS1955: Non-invocable member 'Fixable.DangerousGetPinnableReference' cannot be used like a method.
+                //         fixed (int* p = new Fixable(1))
+                Diagnostic(ErrorCode.ERR_NonInvocableMemberCalled, "new Fixable(1)").WithArguments("Fixable.DangerousGetPinnableReference").WithLocation(6, 25),
+                // (6,25): error CS9365: The given expression cannot be used in a fixed statement
+                //         fixed (int* p = new Fixable(1))
+                Diagnostic(ErrorCode.ERR_ExprCannotBeFixed, "new Fixable(1)").WithLocation(6, 25)
+                );
+        }
+
+        [Fact]
+        public void CustomFixedDelegateErr()
+        {
+            var text = @"
+unsafe class C
+{
+    public static void Main()
+    {
+        fixed (int* p = new Fixable())
+        {
+            System.Console.Write(p[1]);
+        }
+    }
+}
+
+public delegate ref int ReturnsRef();
+
+public struct Fixable
+{
+    public Fixable(int arg){}
+
+    public ReturnsRef DangerousGetPinnableReference => null;
+}
+
+";
+
+            var compVerifier = CreateCompilationWithMscorlib46(text, options: TestOptions.UnsafeReleaseExe);
+
+            compVerifier.VerifyDiagnostics(
+                // (6,25): error CS9365: The given expression cannot be used in a fixed statement
+                //         fixed (int* p = new Fixable())
+                Diagnostic(ErrorCode.ERR_ExprCannotBeFixed, "new Fixable()").WithLocation(6, 25)
+                );
+        }
+
+        #endregion Custom fixed statement tests
```

Comment:
> #endregion Custom fixed statement tests [](start = 7, length = 40)

Can we test the following?
- `fixed (int* p = (dynamic)new Fixable())`

(... I'll expand this list as I get more ideas)

Reply:
Fixed is currently not exposed in semantic model at all, so there is no way to test the differences.
For other suggestions more tests were added.

---
In reply to: [167650001](https://github.com/dotnet/roslyn/pull/24469#discussion_r167650001) [](ancestors = 167650001)